src/ocaml_monadic_ppx.ml

```diff
@@ -1,113 +1,130 @@
+            ([%expr bind [%e bind_expr] [%e cont_function]]
+            [@metaloc expr.pexp_loc])
+        | _ ->
+            (* Nothing left to do.  Just return the body. *)
+            body
+      in
+      bind_wrap value_bindings
+  | Pexp_match (expr_match, cases) ->
+      let f = Ast_helper.Exp.function_ cases in
+      ([%expr bind [%e expr_match] [%e f]] [@metaloc expr.pexp_loc])
+  | Pexp_ifthenelse (expr_if, expr_then, expr_else) ->
+      let expr_else =
+        match expr_else with None -> [%expr ()] | Some case -> case
+      in
+      let cases =
+        [
+          Ast_helper.Exp.case [%pat? true] expr_then;
+          Ast_helper.Exp.case [%pat? false] expr_else;
+        ]
+      in
+      let f = Ast_helper.Exp.function_ cases in
+      ([%expr bind [%e expr_if] [%e f]] [@metaloc expr.pexp_loc])
+  | Pexp_sequence (expr_seq_l, expr_seq_r) ->
+      [%expr bind [%e expr_seq_l] (fun () -> [%e expr_seq_r])]
+      [@metaloc expr.pexp_loc]
+  | _ -> expr
+
+let orzero_expand ~ctxt expr =
+  let loc = Expansion_context.Extension.extension_point_loc ctxt in
+  match expr.pexp_desc with
+  | Pexp_let (Nonrecursive, value_bindings, body) ->
+      (* This is a let%orzero expression.  It's of the form
+           let%orzero $p1 = $e1 and ... and $pn = $en in $e0
+         and we want it to take the form
+           match $e1 with
+           | $p1 -> (match $e2 with
+                     | $p2 -> ...
+                              (match $en with
+                               | $pn -> $e0
+                               | _ -> zero ())
+                     | _ -> zero ())
+           | _ -> zero ()
+      *)
+      let rec orzero_wrap value_bindings' =
+        match value_bindings' with
+        | {
+            pvb_pat = orzero_pattern;
+            pvb_expr = orzero_expr;
+            pvb_attributes = [];
+            pvb_loc = _orzero_loc;
+          }
+          :: value_bindings'' ->
+            (* Recurse and then wrap the resulting body. *)
+            let body' = orzero_wrap value_bindings'' in
+            ([%expr
+               match [%e orzero_expr] with
+               | [%p orzero_pattern] -> [%e body']
+               | _ -> zero ()]
+            [@metaloc expr.pexp_loc])
+        | _ ->
+            (* Nothing left to do.  Just return the body. *)
+            body
+      in
+      orzero_wrap value_bindings
+  | _ -> expr
+
+let bind_extension =
+  Extension.V3.declare "bind" Extension.Context.expression
+    Ast_pattern.(single_expr_payload __)
+    bind_expand
+
+let orzero_extension =
+  Extension.V3.declare "orzero" Extension.Context.expression
+    Ast_pattern.(single_expr_payload __)
+    orzero_expand
+
+let orzero_rule = Ppxlib.Context_free.Rule.extension orzero_extension
+
+let bind_rule = Ppxlib.Context_free.Rule.extension bind_extension
+
+let expr_mapper =
+  object
+    inherit Ast_traverse.map as super
+
+    method! expression e =
+      let e = super#expression e in
```

Comment:
Why do you need to recurse through the AST before the pattern match?

Reply:
I'm sorry but I'm not sure to understand what you mean here. Removing this line produce the following error at compile:

```ocaml
❯ dune build
File "src_test/test.ml", line 150, characters 4-9:
150 |   [%guard y mod 2 == 0];
          ^^^^^
Error: Uninterpreted extension 'guard'.
```

Reply:
Oh, I see, it's because the payload can contain another `%guard` extension node. Thanks!